### PR TITLE
API: Add response filtering with `filter_path` parameter

### LINKED ISSUE #7401: Add `response_transform_script` to allow xpath style selection of response elements
Our JSON responses are currently hard coded. If you only need one piece of information you need to retrieve, and parse, a fairly lengthy JSON document to get it. Although the browser is generally good at this, large responses create problems and require moving data over a, potentially slow, wire, only to throw it away immediately. It would be really nice to only get from the backend (Elasticsearch) the data that is required.

This issue has cropped up in a few places, eg #2149, #7350, #7330 each of which have slightly different requirements.  Instead of supporting multiple options to turn certain parameters on or off, it makes sense to provide a single generic solution that is flexible enough to solve all of these problems.

The best solution that we have found is to use [GPath](http://groovy.codehaus.org/GPath):

> GPath is a path expression language integrated into Groovy which allows parts of nested structured data to be identified. In this sense, it has similar aims and scope as XPath does for XML.

We will add a `response_transform_script` parameter on all APIs (in the body or in the query string) which provides the full JSON response as a GPath object (eg `_response`) which can be manipulated at will. For instance, to return just the  `_source` fields from the `hits` array from a search request, you could do:

    GET /_search
    {
        "response_transform_script: {
            "script": "_json.hits.hits.collect { it._source }"
        }
    }

### PR DESCRIPTION
**Pull request by tlrx**
This change adds a new "_path" parameter that can be used to filter and reduce the responses returned by the REST API of elasticsearch.

For example, returning only the shards that failed to be optimized:

```
curl -XPOST 'localhost:9200/beer/_optimize?_path=_shards.failed'
{"_shards":{"failed":0}}%
```

It supports multiple filters (separated by a comma):

```
curl -XGET 'localhost:9200/_mapping?pretty&_path=*.mappings.*.properties.name,*.mappings.*.properties.title'
```

It also supports the YAML response format. Here it returns only the `_id` field of a newly indexed document:

```
curl -XPOST 'localhost:9200/library/book?_path=_id' -d '---hello:\n  world: 1\n'

---
_id: "AU0j64-b-stVfkvus5-A"
```

It also supports wildcards. Here it returns only the host name of every nodes in the cluster:

```
curl -XGET 'http://localhost:9200/_nodes/stats?_path=nodes.*.host*'
{"nodes":{"lvJHed8uQQu4brS-SXKsNA":{"host":"portable"}}}
```

And "**" can be used to include sub fields without knowing the exact path. Here it returns only the Lucene version of every segment:

```
curl 'http://localhost:9200/_segments?pretty&_path=indices.**.version'
{
  "indices" : {
    "beer" : {
      "shards" : {
        "0" : [ {
          "segments" : {
            "_0" : {
              "version" : "5.2.0"
            },
            "_1" : {
              "version" : "5.2.0"
            }
          }
        } ]
      }
    }
  }
}
```

Note that elasticsearch sometimes returns directly the raw value of a field, like the "_source" field. If you want to filter the response that include _source fields like in Search or Get responses , you should consider using the already existing "fields" parameter:

```
 curl -XGET 'localhost:9200/beer/tasty/3672?pretty&fields=category'
 {
   "_index" : "beer",
   "_type" : "tasty",
   "_id" : "3672",
   "_version" : 1,
   "found" : true,
   "fields" : {
     "category" : [ "German Lager" ]
   }
 }
```

The "_path" parameter can be used to further reduce the result:

```
curl -XGET 'localhost:9200/beer/tasty/3672?pretty&fields=category&_path=fields.category'
{
  "fields" : {
    "category" : [ "German Lager" ]
  }
}

```

Closes #7401

### CODE REVIEW

**Comment by tlrx** [comment]
Note: 
* if the provided filter specified in `_path` does not match something, an empty response is returned
* if the API returns an error, the error is not filtered even if the `_path` is specified

**Comment by clintongormley** [comment]
> if the provided filter specified in _path does not match something, an empty response is returned

We should return an empty JSON object rather than an empty string, otherwise the client needs to check whether JSON has been returned.

**Comment by tlrx** [comment]
@clintongormley thanks, I updated the code to reflect your last comment.

**Comment by tlrx** [comment]
@spinscale Can you please have a look? Thanks

File: src/main/java/org/elasticsearch/common/xcontent/json/JsonXContent.java [hunk 1]
```diff
@@ -63,14 +64,27 @@ public byte streamSeparator() {
...
+
     @Override
     public XContentGenerator createGenerator(OutputStream os) throws IOException {
-        return new JsonXContentGenerator(jsonFactory.createGenerator(os, JsonEncoding.UTF8));
+        return newXContentGenerator(jsonFactory.createGenerator(os, JsonEncoding.UTF8));
+    }
+
+    @Override
+    public XContentGenerator createGenerator(OutputStream os, String[] filters) throws IOException {
+        if ((filters == null) || (filters.length == 0)) {
```
Comment (spinscale) [inline src/main/java/org/elasticsearch/common/xcontent/json/JsonXContent.java]: use `CollectionUtils.isEmpty()`?

File: src/main/java/org/elasticsearch/common/xcontent/smile/SmileXContent.java [hunk 2]
```diff
@@ -62,14 +63,27 @@ public byte streamSeparator() {
...
+
     @Override
     public XContentGenerator createGenerator(OutputStream os) throws IOException {
-        return new SmileXContentGenerator(smileFactory.createGenerator(os, JsonEncoding.UTF8));
+        return newXContentGenerator(smileFactory.createGenerator(os, JsonEncoding.UTF8));
+    }
+
+    @Override
+    public XContentGenerator createGenerator(OutputStream os, String[] filters) throws IOException {
+        if ((filters == null) || (filters.length == 0)) {
```
Comment (spinscale) [inline src/main/java/org/elasticsearch/common/xcontent/smile/SmileXContent.java]: use `CollectionUtils.isEmpty()`?

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    protected void assertBinary(String expected, XContentBuilder builder) {
+        assertNotNull(builder);
+        assertNotNull(expected);
+        assertNotNull(expected, not(isEmptyString()));
+
+        try {
+            XContentParser jsonParser = XContentFactory.xContent(XContentType.JSON).createParser(expected.getBytes("UTF8"));
+            XContentParser xsonParser = XContentFactory.xContent(builder.contentType()).createParser(builder.bytes());
```
Comment (spinscale) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java]: maybe different name?

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                XContentParser.Token token1 = jsonParser.nextToken();
+                XContentParser.Token token2 = xsonParser.nextToken();
+                if (token1 == null) {
+                    assertThat(token2, nullValue());
+                    return;
+                }
+                assertThat(token1, Matchers.equalTo(token2));
+                switch (token1) {
+                    case FIELD_NAME:
+                        assertThat(jsonParser.currentName(), Matchers.equalTo(xsonParser.currentName()));
```
Comment (spinscale) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java]: static import of Matchers or just static import `Matchers.is()` (but thats my personal preference over `equalTo`...)

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    protected abstract String resultSimpleArrayOfObjectsProperty();
+
+    @Test
+    public void testRecurseField1() throws Exception {
+        assertXContentBuilder(resultRecurseField1(), sample("**.name"));
+    }
+
+    protected abstract String resultRecurseField1();
```
Comment (spinscale) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java]: why cant you use the xcontent parsers to produce those, based on `getXContentType()`, then there is no need to do this for each xcontent... guess I am missing something here?

**Comment by spinscale** [comment]
A couple of things

First, I really would like to see some simple benchmarking, so we have some ballpark numbers, how this affects performance. Especially because `write` operations that create sub-hierarchies also create new objects.

Minor thing: `FilteringJsonGenerator.parse()` - why not use an array instead of XContentFilter and `next` (like your own custom linked list).

One other thing, I know that jackson has some filtering capabiltiies built-in, I have no ideas if we can make use of them though, see [here](http://wiki.fasterxml.com/JacksonFeatureJsonFilter), [here](http://www.cowtowncoder.com/blog/archives/2011/09/entry_461.html) and [here](http://www.cowtowncoder.com/blog/archives/2011/02/entry_443.html)

**Comment by tlrx** [comment]
> One other thing, I know that jackson has some filtering capabiltiies built-in, I have no ideas if we can make use of them though, see here, here and here

That's the first thing we checked before starting working on this pull request. The current filtering capabilities of Jackson (version <2.5.1) can be used only with jackson-databinding lib and apply only to POJOs but not when you use jackson to generate JSON in streaming mode. 

Jackson 2.6 will integrate an interesting filtering feature that will work with JSON streaming (see release note [here](https://github.com/FasterXML/jackson-core/blob/master/release-notes/VERSION#L22) and test class [here](https://github.com/FasterXML/jackson-core/blob/master/src/test/java/com/fasterxml/jackson/core/filter/JsonPointerGeneratorFilteringTest.java)). It currently uses [JSONPointer](https://tools.ietf.org/html/rfc6901) to filter properties but I think we will be able to implement our own [TokenFilter](https://github.com/FasterXML/jackson-core/blob/master/src/main/java/com/fasterxml/jackson/core/filter/TokenFilter.java) later, once version 2.6 is released.

File: src/main/java/org/elasticsearch/common/xcontent/json/JsonXContent.java (hunk 1 above)
Comment (tlrx) [inline src/main/java/org/elasticsearch/common/xcontent/json/JsonXContent.java]: Ok

File: src/main/java/org/elasticsearch/common/xcontent/smile/SmileXContent.java (hunk 2 above)
Comment (tlrx) [inline src/main/java/org/elasticsearch/common/xcontent/smile/SmileXContent.java]: Ok

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java (hunk 3 above)
Comment (tlrx) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java]: Ok, changed for testParser

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java (hunk 4 above)
Comment (tlrx) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java]: Ok

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java (hunk 5 above)
Comment (tlrx) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTests.java]: You're not missing something, I did it in the `testRawField()` method but I can do it for all tests.

**Comment by tlrx** [comment]
> First, I really would like to see some simple benchmarking, so we have some ballpark numbers, how this affects performance. Especially because write operations that create sub-hierarchies also create new objects.

I added the `FilteringJsonGeneratorBenchmark` and put some numbers [here](https://gist.github.com/tlrx/960ce767ffc084dee2d6) (not very readable - I apologize for that). 

There are no real surprise in the benchmark: 
- when writing a small number of documents (less than 100) with a small number of fields each(1-10), the filtering adds an overhead that make it less efficient than the current version
 
- when writing a larger amount of data (ie [large number of small docs](https://gist.github.com/tlrx/960ce767ffc084dee2d6#file-filteringjsongeneratorbenchmark-L64-L74), [small number of large docs](https://gist.github.com/tlrx/960ce767ffc084dee2d6#file-filteringjsongeneratorbenchmark-L240-L250)) there's a point where it becomes more efficient to check if fields must be included rather than writing more data. This point varies depending on the document size and the selectivity of the filter, but in this test it is in between 1-10%.
 

> Minor thing: FilteringJsonGenerator.parse() - why not use an array instead of XContentFilter and next (like your own custom linked list).

There are room for improvements like this one (I'm also thinking of not creating FilterContext for sub fields when a parent field matches the end of a filter), but it might complexify a the code and I'm not sure it will be more efficient.

I'm currently tryong some of these improvement to see if they are really pertinent or not.

**Comment by tlrx** [comment]
Thinking of improvements again, we should be able to reuse the objects instead of creating new ones for every sub fields. I'll try to improve that.

**Comment by clintongormley** [comment]
> when writing a small number of documents (less than 100) with a small number of fields each(1-10), the filtering adds an overhead that makes it less efficient than what we have today

I assume that the code path (and performance) remains the same as today if no `_path` is specified?

**Comment by tlrx** [comment]
@clintongormley yes. Branching is done [here](https://github.com/elastic/elasticsearch/pull/10980/files#diff-dea72b6ed4d9da9b9008b0e185d77887R77) where we fall back to the normal XContent generator.

File: src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilterContext.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+     */
+    private String property;
+
+    /**
+     * List of XContentFilter matched by the current filtering context
+     */
+    private List<XContentFilter> matchings;
+
+    /**
+     * Flag to indicate if the field/property must be write
```
Comment (spinscale) [inline src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilterContext.java]: ... written

File: src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGenerator.java [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+                ctx.addMatching(matching);
+            }
+        }
+        return ctx;
+    }
+
+    /**
+     * Adds a context instance to the pool in order to reuse it if needed
+     */
+    private void release(FilterContext ctx) {
```
Comment (spinscale) [inline src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGenerator.java]: was puzzled about the naming of `release` here for a second for adding a filter context.
Also, the below code, just means, this context can never be reused by not being added to the queue, right? (Just want to make sure I get this right)

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGeneratorBenchmark.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.IOException;
+import java.util.Arrays;
+
+/**
+ * Benchmark class to compare filtered and unfiltered XContent generators.
+ */
+public class FilteringJsonGeneratorBenchmark {
+
+    public static void main(String[] args) throws IOException {
+        final XContent XCONTENT = (XContent) JsonXContent.jsonXContent;
```
Comment (spinscale) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGeneratorBenchmark.java]: unneeded cast?

**Comment by spinscale** [comment]
added some absolutely minor comments...

First, this doesnt compile using mvn, due to the benchmark class and some locale, but not really an issue.

Few last things:
* the kind-of linkedlist data structure in the FilterContext, which could possibly be replaced by a simple array?
* This seems the only part, where we use `**` as 'greedy' wildcard in Elasticsearch, not sure if it is a problem, just to be aware
* No documentation so far?

Maybe we should also note somewhere to switch to the jackson filter features with the 2.6 release, I'd hate to forget that...

LGTM apart from that

**Comment by rjernst** [comment]
I'm late to the party here, but I have a simple question. Why is the parameter `_path` and not `path`? Why use the underscore when we control all the url params anyways? And all the other examples I know of (eg `fields`, `pretty`, `human`) do not use underscore?

**Comment by clintongormley** [comment]
@rjernst only because it works well with the existing `_source` filtering.  I kinda like the underscore because it is a "global" param (mind you, so is pretty).

**Comment by clintongormley** [comment]
> This seems the only part, where we use ** as 'greedy' wildcard in Elasticsearch, not sure if it is a problem, just to be aware

Elsewhere we use `*` as a greedy wildcard, which I think is problematic.  We took the `**` syntax from ant, and I'd actually be tempted to reuse it elsewhere.

**Comment by kimchy** [comment]
actually, since this is a "rest" level parameter, and for those, we have `pretty` and `human`, I think we should just call it `path`, or maybe be more explicit and call it `filter_path`

**Comment by rjernst** [comment]
+1 on `filter_path`

**Comment by kimchy** [comment]
here is another suggestion, calling it `filter_path_include`, in the future we might want to support `filter_path_exclude`. 

The reason I am raising it is that today, we support source level include/exclude when fetching source documents. With this infrastructure, if we eventually expand it to support include*s*/exclude*s*, we can also support zero copy (as in, not create a map of maps/lists representation of source) when someone asks for source includes/includes (we can just create a bytes based filtering generator, and copy structure from the parser right into it). 

I helped a fellow on IRC today where the loading and parsing of source include was the perf bottleneck (from 1ms it got up to 170ms for ~1000 docs). I think this will help a lot there eventually.

**Comment by nik9000** [comment]
Sorry I just jumped into this late. It makes me think of jq. Nothing else
to add.

**Comment by clintongormley** [comment]
What about using a `-` prefix for excludes as in `"*.foo,-*.foo.bar"`. This is what is used by the (non-streaming) AntPath filter for Jackson https://github.com/Antibrumm/jackson-antpathfilter#usage

File: src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGenerator.java (hunk 7 above)
Comment (tlrx) [inline src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGenerator.java]: I renamed `acquire` into `get` and `release` into `put`, does it sound better?

Right, we only keep a limited number of contexts (10) around in order to reuse them. In my tests this list did not exceed 3 instances.

File: src/test/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGeneratorBenchmark.java (hunk 8 above)
Comment (tlrx) [inline src/test/java/org/elasticsearch/common/xcontent/support/filtering/FilteringJsonGeneratorBenchmark.java]: Yes, thanks

**Comment by tlrx** [comment]
@kimchy now I'm reading your comment again, I think that's a neat idea :)

@clintongormley I like the `+/-` notation but this is not consistent with the `_source?_source_include=*.id&_source_exclude=entities` notation... I think we should use the same notation (whatever it is) in both endpoints.

**Comment by tlrx** [comment]
@spinscale I rebased and updated the code according to your last comments. Can you have a look and do some manual testing please? Thanks a lot :)

Note: latest benchmark numbers are [here](https://gist.github.com/tlrx/65d2a46e394b13aaa11e).

**Comment by tlrx** [comment]
@spinscale I just added the REST tests. The parameter is still named `_path` until we reach an agreement on the final name.

File: rest-api-spec/api/nodes.stats.json [hunk 9]
```diff
@@ -56,6 +56,10 @@
           "options" : ["node", "indices", "shards"],
           "default" : "node"
         },
+        "_path": {
```
Comment (spinscale) [inline rest-api-spec/api/nodes.stats.json]: I think the `_path` needs to be added to all the methods supporting it, to make sure the clients are aware

File: docs/reference/api-conventions.asciidoc [hunk 10]
```diff
@@ -82,6 +82,96 @@ consumption.  The default for the `human` flag is
...
+      }
+    }
+  }
+}
+--------------------------------------------------
+
+Note that elasticsearch sometimes returns directly the raw value of a field,
+like the `_source` field. If you want to filter the response that include
+_source fields, you should consider using the already existing `_source`
+parameter (see <<get-source-filtering,Get API>> for more details).
```
Comment (spinscale) [inline docs/reference/api-conventions.asciidoc]: maybe mention in an example, like filtering for `hits.hits._source.name` will not work and return an empty result

**Comment by spinscale** [comment]
left one comment, apart from that it feels like we are close to getting this in, when we get the naming stuff resolved... /cc @clintongormley

**Comment by clintongormley** [comment]
The `_source` parameter is essentially a synonym for `_source_include`, so I'd go with `filter_path` for now.  We can always add `filter_path_include|exclude` later on.

**Comment by tlrx** [comment]
@clintongormley sounds good to me, thanks

File: rest-api-spec/api/nodes.stats.json (hunk 9 above)
Comment (clintongormley) [inline rest-api-spec/api/nodes.stats.json]: No, like pretty and `_source`, we can just support this parameter everywhere, so don't add it to the rest spec.

**Comment by tlrx** [comment]
@spinscale Updated with your last comment :)

**Comment by clintongormley** [comment]
@spinscale @kimchy does this need another review, or are we good to go?

**Comment by spinscale** [comment]
LGTM on my side

**Comment by clintongormley** [comment]
w00t

**Comment by rashidkpc** [comment]
WIN Thanks so much everyone for making this happen. This is huge for Kibana!

**Comment by dkroehan** [comment]
Is this already supported via the Java API?

**Comment by tlrx** [comment]
@dkroehan The `filter_path` parameter is used as the REST level (like the `pretty` parameter) to render JSON responses and therefore it is not part of the Java API.

If you need to render `XContent` object instances using filtering in Java, you can use the [XContent.builder(XContent xContent, String[] filters)](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java#L85) method

**Comment by nmors** [comment]
are excludes supported yet, or just includes?

**Comment by tlrx** [comment]
@nmors Using the REST API and the `filter_path` parameter, only includes are supported. Using `XContent` filtering in Java, excludes are supported but you can't mix them both.

**Comment by nmors** [comment]
Thanks, I'm actually using the **javascript** client. I'm using the **filterPath** parameter like so (I guess it's a wrapper for filter_path), all is working fine! It would be good to be able to use excludes though, can be very helpful for me as it is web traffic that I would like to keep small. I've managed to cut the response size in half by using the following:

```
  index: my_index,
  type: my_type,
  body: q,
  filterPath: [
  	"aggregations.my_agg.buckets.**.value",
  	"aggregations.my_agg.buckets.**.key"
  ]
```

**Pull request merged**